File: csharp_component/component/hub/centerproxy.cs
using System;
using System.Threading.Tasks;

namespace Hub
{
	public class CenterProxy
    {
        private readonly Abelkhan.center_caller _center_caller;

        public CenterProxy(Abelkhan.Ichannel ch)
		{
			is_reg_center_sucess = false;
            _ch = ch;
            _center_caller = new Abelkhan.center_caller(ch, Abelkhan.ModuleMgrHandle._modulemng);
        }

        public void reg_hub(Action callback)
        {
            Log.Log.trace("begin connect center server");

            _center_caller.reg_server_mq("hub", Hub.type, Hub.name).callBack(() =>
            {
                callback.Invoke();
                Log.Log.trace("connect center server sucessed");
            }, () =>
            {
                Log.Log.trace("connect center server faild");
            }).timeout(5000, () =>
            {
                Log.Log.trace("connect center server timeout");
            });
        }

        public Task<bool> reconn_reg_hub()
        {
            Log.Log.trace("begin connect center server");

            var task_ret = new TaskCompletionSource<bool>();

            _center_caller.reconn_reg_server_mq("hub", Hub.type, Hub.name).callBack(() => {
                Log.Log.trace("reconnect center server sucessed");
                task_ret.SetResult(true);
            }, () => {
                Log.Log.err("reconnect center server faild");
                task_ret.SetResult(false);
            }).timeout(5000, () => {
                Log.Log.err("reconnect center server timeout");
                task_ret.SetResult(false);
            });

            return task_ret.Task;
        }

        public void heartbeat()
        {
            _center_caller.heartbeat(Hub.tick).callBack(() => {
                Log.Log.trace("heartbeat center server sucessed");

                timetmp = Service.Timerservice.Tick;

            }, () => {
                Log.Log.err("heartbeat center server faild");
            }).timeout(5000, () => {
                Log.Log.err("heartbeat center server timeout");
            });
            Log.Log.trace("begin heartbeath center server tick:{0}!", Service.Timerservice.Tick);
        }

        public void closed()
        {
            _center_caller.closed();
        }

        public long timetmp = Service.Timerservice.Tick;
        public bool is_reg_center_sucess;
        public readonly Abelkhan.Ichannel _ch;

    }
}

File: py_hub/centerproxy.py
import abelkhan
import center
from collections.abc import Callable
import hub_server

class centerproxy(object):
    def __init__(self, _ch:abelkhan.Ichannel, _modulemng:abelkhan.modulemng, _hub:hub_server.hub) -> None:
        self.timetmp = abelkhan.timetmp()
        self.is_reg_center_sucess = False
        self.ch = _ch
        self.center_caller = center.center_caller(self.ch, _modulemng)
        self.hub = _hub

    def __refresh_timetmp__(self):
        self.timetmp = abelkhan.timetmp()

    def reg_hub(self, callback:Callable[[]]) -> None:
        print("begin connect center server")

        self.center_caller.reg_server_mq("hub", self.hub.type, self.hub.name).callBack(
            callback,
            lambda : print("connect center server faild")
        ).timeout(5000, lambda : print("connect center server timeout"))

    def reconn_reg_hub(self, callback:Callable[[]]) -> None:
        print("begin connect center server")

        self.center_caller.reconn_reg_server_mq("hub", self.hub.type, self.hub.name).callBack(
            callback,
            lambda : print("reconnect center server faild")
        ).timeout(5000, lambda : print("reconnect center server timeout"))

    def heartbeat(self) -> None:
        print("begin heartbeath center server tick:!", abelkhan.timetmp())

        self.center_caller.heartbeat(abelkhan.timetmp()).callBack(
            lambda : self.__refresh_timetmp__(),
            lambda : print("heartbeat center server faild")
        ).timeout(5000, lambda : print("heartbeat center server timeout"))

    def closed(self):
        self.center_caller.closed()